Gallery/Contours/NCL_h_lat_6.py

```diff
@@ -104,12 +105,17 @@
                          labelfontsize=16)
 
 # Create second y-axis to show geo-potential height.
-# Currently we're using arbitrary values for height as we haven't figured out
-# how to make this work properly yet
 axRHS = ax.twinx()
 
+# Use MetPy's pressure_to_height_std function to get standard atmosphere height conversion
+heights = mpcalc.pressure_to_height_std(ds.lev).values
+min_height = min(heights)
+max_height = max(heights)
+
 # Use geocat.viz.util convenience function to set axes tick values
-gv.set_axes_limits_and_ticks(axRHS, ylim=(0, 32), yticks=np.arange(4, 32, 4))
+gv.set_axes_limits_and_ticks(axRHS,
+                             ylim=(min_height, max_height),
+                             yticks=np.arange(4, max_height, 4))
```

Comment:
Wait, here shouldn't we be doing something like getting the positions (in height) we want to label and then labeling those positions? Definitely correct me if I'm reading this wrong but won't this just evenly space some ticks between the max and min position?

Reply:
You are correct that this will evenly space ticks, but now those ticks and their labels are at an accurate location corresponding to the left vertical axis (pressure). Whereas before, they were arbitrary numbers that did not match the pressure values.

The left ticks are not in identical spots to the right ticks though, but I don't think that is what the user would want and the numbers wouldn't be as pretty.

Reply:
I think if the user has specific height values they want to plot (as opposed to integers within the height range at set intervals), it would be pretty straightforward for them to use a numpy array with those values for yticks, as long as min_height and max_height are calculated from pressure.

Reply:
Or maybe I have to think about this longer

Reply:
I think I'm confused about the positions of the ticks between the min and max values. I don't think we can assume that they'll be evenly spaced, right?

Reply:
https://www.ncl.ucar.edu/Applications/Images/h_lat_6_lg.png Do these look evenly spaced to you? They look slightly off. Well it is at least closer to true than what we had -- but I think you're right. Now the top and bottom are pinned correctly, but the heights are not necessarily 1:1 along the axis.

Reply:
Also, I'm a little concerned that axis on the new plot looks pretty different from the original ncl.

Reply:
> https://www.ncl.ucar.edu/Applications/Images/h_lat_6_lg.png Do these look evenly spaced to you? They look slightly off. Well it is at least closer to true than what we had -- but I think you're right. Now the top and bottom are pinned correctly, but the heights are not necessarily 1:1 along the axis.

Yeah, I don't think those are evenly spaced, but just barely. Especially the top two tick marks. I'm assuming this was to get nice, evenly spaced values for the labels?

Reply:
If I change the ticks to 

```
# Create second y-axis to show geo-potential height.
axRHS = ax.twinx()

# Use MetPy's pressure_to_height_std function to get standard atmosphere height conversion
heights = mpcalc.pressure_to_height_std(ds.lev).values
min_height = min(heights)
max_height = max(heights)

axRHS.set_ylim(min_height, max_height)
axRHS.set_yticks(heights)
```

I get 
![image](https://user-images.githubusercontent.com/46687291/230679121-cd4e52b1-bc5c-4012-bab5-88869ed84ba5.png)
is this closer? The numbers aren't pretty, but are they at least at the correct spaces?

Reply:
In this [stackoverflow question](https://stackoverflow.com/questions/14153816/matplotlib-axis-labels-how-to-find-out-where-they-will-be-located) people seem to imply that altitude is linear, but I don't think they're right.

Reply:
NCL seems to do the conversion here: https://github.com/NCAR/ncl/blob/8a96101fe14d0cf0f5ed66a5e6b1733084bc69df/ni/src/examples/gsun/gsn_csm.ncl#L2659

I will try to recreate this in Python

Reply:
So, I think the thing to do might be to think of this the other way around. 

What if we set both y-axes to be the same scale. Then, use `height_to_pressure_std` instead of `pressure_to_height_std` to get the pressure values that correspond to `np.arange(4,24,4)` (or something like that). Then, set the RHS tick values to be the corresponding calculated pressures and set the labels to be the height values.

This is convoluted enough it might make sense to make this into a geocat-viz function

Reply:
![image](https://user-images.githubusercontent.com/46687291/230685929-5bbee0d0-89ea-48bd-872f-a19698a4a70c.png)

I followed the NCL steps and produced code that should probably be in its own function to be reused in other plots that have altitude and pressure plotted. It still doesn't match the NCL so I'd love to confer with someone on the math next week:

```
# Create second y-axis to show geo-potential height.
axRHS = ax.twinx()

# Use MetPy's pressure_to_height_std function to get standard atmosphere height conversion
heights = mpcalc.pressure_to_height_std(ds.lev).values

min_height = min(heights)
max_height = max(heights)

# Auto select "nice" heights based on range
height_range = max_height - min_height
if (height_range < 35):
    step = 4
else:
    if (height_range < 70):
        step = 7
    else:
        step = 10
  
heights_nice = np.arange(int(min_height), int(max_height), step)

# Use a cubic spline interpolation to send "nice" height values back to pressure tick locations
cubic_spline = scipy.interpolate.CubicSpline(heights, np.flip(ds.lev))
pressures_nice = cubic_spline(heights_nice)

gv.set_axes_limits_and_ticks(axRHS,
                            ylim=ax.get_ylim()[::-1],
                            yticks=pressures_nice,
                            yticklabels=heights_nice)
axRHS.tick_params(labelsize=12)  # manually set tick label size
```

Reply:
> ght be to think of this the other way around.

Just saw this. I like this idea. We might still need to do both conversions to get our "nice" heights range, and we can make this a fx where the user inputs their step size

Reply:
> Just saw this. I like this idea. We might still need to do both conversions to get our "nice" heights range, and we can make this a fx where the user inputs their step size

Yeah! I imagine the user could provide an array of their intended tick locations as an `np.arange` or list or something

Reply:
Okay trying your suggestion (I like the idea of using metpy again instead of scipy) I have the following code:
```
# Create second y-axis to show geo-potential height.
axRHS = ax.twinx()

# Use MetPy's pressure_to_height_std function to get standard atmosphere height conversion
heights = mpcalc.pressure_to_height_std(ds.lev).values
heights_nice = np.arange(int(min(heights)), int(max(heights)), 4)

# Send "nice" height values back to pressure tick locations
pressures_nice = mpcalc.height_to_pressure_std(heights_nice * units('km')).magnitude

gv.set_axes_limits_and_ticks(axRHS,
                            ylim=ax.get_ylim(),
                            yticks=pressures_nice,
                            yticklabels=heights_nice)
```
It is flipped somehow, which I keep going in circles trying to change
![image](https://user-images.githubusercontent.com/46687291/230689450-3541a4dc-843d-4925-8b39-2390191240eb.png)

But I will keep working on this next week.

Reply:
The biggest problem was that I had to set the second axis to log as well (even though the plt as a whole was set to it).

It ended up only being a few lines to get the "nice" height pressure values and setting it to log. So I'm not sure if we need a new function at this stage.

Reply:
![image](https://user-images.githubusercontent.com/46687291/230743059-c8fd0eb9-f50a-4402-9dcc-98cb62f57a29.png)
Just saw that I didn't post the new image with the logarithmic right y-axis